server/src/main/java/io/druid/server/http/MetadataResource.java

```diff
@@ -82,103 +82,81 @@ public MetadataResource(
-                  {
-                    Resource resource = new Resource(input, ResourceType.DATASOURCE);
-                    Action action = Action.READ;
-                    Pair<Resource, Action> key = new Pair<>(resource, action);
-                    if (resourceAccessMap.containsKey(key)) {
-                      return resourceAccessMap.get(key).isAllowed();
-                    } else {
-                      Access access = authorizationInfo.isAuthorized(key.lhs, key.rhs);
-                      resourceAccessMap.put(key, access);
-                      return access.isAllowed();
-                    }
-                  }
-                }
-            )).build();
-      } else {
-        druidDataSources =
-            Collections2.filter(
-                metadataSegmentManager.getInventory(),
-                new Predicate<DruidDataSource>()
+      dataSourceNamesPostAuth = ImmutableSet.copyOf(
+          Sets.filter(
+              dataSourceNamesPreAuth,
+              new Predicate<String>()
+              {
+                @Override
+                public boolean apply(String input)
                 {
-                  @Override
-                  public boolean apply(DruidDataSource input)
-                  {
-                    Resource resource = new Resource(input.getName(), ResourceType.DATASOURCE);
-                    Action action = Action.READ;
-                    Pair<Resource, Action> key = new Pair<>(resource, action);
-                    if (resourceAccessMap.containsKey(key)) {
-                      return resourceAccessMap.get(key).isAllowed();
-                    } else {
-                      Access access = authorizationInfo.isAuthorized(key.lhs, key.rhs);
-                      resourceAccessMap.put(key, access);
-                      return access.isAllowed();
-                    }
+                  Resource resource = new Resource(input, ResourceType.DATASOURCE);
+                  Action action = Action.READ;
+                  Pair<Resource, Action> key = new Pair<>(resource, action);
+                  if (resourceAccessMap.containsKey(key)) {
+                    return resourceAccessMap.get(key).isAllowed();
+                  } else {
+                    Access access = authorizationInfo.isAuthorized(key.lhs, key.rhs);
+                    resourceAccessMap.put(key, access);
+                    return access.isAllowed();
                   }
                 }
-            );
-      }
+              }
+          )
+      );
     } else {
-      druidDataSources = metadataSegmentManager.getInventory();
+      dataSourceNamesPostAuth = dataSourceNamesPreAuth;
     }
 
-    if (includeDisabled != null) {
-      return builder.entity(
-          Collections2.transform(
-              druidDataSources,
-              new Function<DruidDataSource, String>()
+    // Cannot do both includeDisabled and full, let includeDisabled take priority
```

Comment:
https://github.com/druid-io/druid/pull/3047 fixes it without any such restrictions. Although I am ok with this also if others do not have any problems with it.